### PR TITLE
Wrap cusparseSpMV

### PR DESCRIPTION
**Pull request by marius311**
This wraps the new `cusparseSpMV` which does `sparse matrix * vector`, and should fix some of https://github.com/JuliaGPU/CUDA.jl/issues/322. Some of the code here would also make it pretty trivial to wrap other generic cuSparse API functions. 

I'm really a newbie here, so I'm opening this mainly as a way to get some help for now. I think most of this is right and possibly very close, but I'm still getting the following error from the API call to `cusparseSpMV`

```julia

julia> using CUDA, CUDA.CUSPARSE, SparseArrays

julia> A = CuSparseMatrixCSR(sprand(Float32,10,10,0.2))

julia> v = cu(rand(10))

julia> A * v
ERROR: CUSPARSEError: initialization error (code 1, CUSPARSE_STATUS_NOT_INITIALIZED)
Stacktrace:
 [1] throw_api_error(::CUDA.CUSPARSE.cusparseStatus_t) at /home/marius/work/baylens/dev/CUDA/lib/cusparse/error.jl:21
 [2] macro expansion at /home/marius/work/baylens/dev/CUDA/lib/cusparse/error.jl:32 [inlined]
 [3] cusparseSpMV(::Ptr{Nothing}, ::Char, ::Array{Float32,1}, ::CUDA.CUSPARSE.CuSparseMatrixDescriptor, ::CUDA.CUSPARSE.CuDenseVectorDescriptor, ::Array{Float32,1}, ::CUDA.CUSPARSE.CuDenseVectorDescriptor, ::Type{T} where T, ::CUDA.CUSPARSE.cusparseSpMVAlg_t, ::CuPtr{Nothing}) at /home/marius/work/baylens/dev/CUDA/lib/utils/call.jl:93
 [4] mv!(::Char, ::Float32, ::CuSparseMatrixCSR{Float32}, ::CuArray{Float32,1}, ::Float32, ::CuArray{Float32,1}, ::Char) at /home/marius/work/baylens/dev/CUDA/lib/cusparse/wrappers.jl:984
 [5] mul! at /home/marius/work/baylens/dev/CUDA/lib/cusparse/interfaces.jl:12 [inlined]
 [6] *(::CuSparseMatrixCSR{Float32}, ::CuArray{Float32,1}) at /home/marius/src/julia-1.5/usr/share/julia/stdlib/v1.5/LinearAlgebra/src/matmul.jl:51
 [7] top-level scope at REPL[3]:1
 [8] include_string(::Function, ::Module, ::String, ::String) at ./loading.jl:1088
```

How do I go about debugging an error like this?

### CODE REVIEW

**Comment by marius311** [comment]
Ok, got `A*v` and `A'*v` working for CSR. I wasn't quite doing the matrix descriptor pointers right. Running on a different machine gave me a different error which was more helpful. 

I added tests and I think its ready for review. Note I couldn't get the tests to actually run on any machine even before this PR, so I haven't run them except by hand. Probably need your help verifying they're right.

File: lib/cusparse/helpers.jl [hunk 1]
```diff
@@ -1,9 +1,26 @@
...
+Base.unsafe_convert(::Type{cusparseDnVecDescr_t}, desc::CuDenseVectorDescriptor) = desc.handle
+
+
+## matrix descriptor
+
+mutable struct CuMatrixDescriptor
     handle::cusparseMatDescr_t
 
-    function CuSparseMatrixDescriptor()
+    function CuMatrixDescriptor()
```
Comment (marius311) [inline lib/cusparse/helpers.jl:9]: Note the name change from CuSparseMatrixDescriptor to match the CUDA API `cusparseMatDescr_t` and `cusparseSpMatDescr_t`.

**Comment by maleadt** [comment]
Thanks, I had some local changes already which I'll push here tomorrow or next week. I first wanted to understand the API a little better, right now it looks like CUSPARSE has two APIs (the BLAS-like level 1-2-3 one + extensions, and a new one with more generic operations), each of which with its own descriptors. But I'm not sure whether that's entirely accurate.

**Comment by marius311** [comment]
Thanks, yea I saw in another thread you mentioned you're working on it but I just needed this myself quickly so I went ahead anyway. Feel free to merge or start fresh with your changes.

Re: the API, that sounds about right. Most of the functionality seems to exist in both APIs, except, as I think you know, the CSR `mv` dissapeared from the old one, so it might make sense to just get this in CUDA.jl to fix the regression before fully wrapping everything else.

**Comment by maleadt** [comment]
bors try

**Comment by maleadt** [comment]
bors try

The reason for the delay is that I was considering reworking the sparse vector/matrix structs to each have their own descriptors instead of continuously recreating them (although we'd have to do that anyway to get a dense descriptor from regular CuArrays). Haven't got the time to figure out a good design yet, though.

**Comment by marius311** [comment]
Sorry, test failures were stupid mistake on my part, tests pass locally now. 

> The reason for the delay is that I was considering reworking the sparse vector/matrix structs to each have their own descriptors instead of continuously recreating them (although we'd have to do that anyway to get a dense descriptor from regular CuArrays). Haven't got the time to figure out a good design yet, though.

No worries, thanks for letting me know. After this PR is merged I was thinking of filling in some more functionality for COO matrices based on https://github.com/marius311/CUDA.jl/commit/c5214df4378b4d9f9be81053b93ccdfb9f288ba4 and submitting another PR. Would it be better to wait until you decide what to do about these descriptors?

**Comment by maleadt** [comment]
> Would it be better to wait until you decide what to do about these descriptors?

Not necessarily, we can always rework that afterwards. We'll always have a `CuSparseMatrixCOO` type, so it shouldn't affect user code.
bors try

**Comment by maleadt** [comment]
Maybe this is a good time to drop support for CUDA 9.x and 10.0, for which NVIDIA doesn't distribute cuDNN 8 either. Before that, let's maybe have one more release that brings multi-device capabilities to those users, so let's wait to merge this a little longer.

**Comment by maleadt** [comment]
bors try

**Comment by maleadt** [comment]
Thanks for the PR!

**Pull request merged**